.github/workflows/main.yml

```diff
@@ -361,6 +361,56 @@ jobs:
           name: 'obs-studio-windows-${{ matrix.arch }}-${{ steps.setup.outputs.commitHash }}'
           path: '${{ env.FILE_NAME }}'
 
+  source_build:
+    name: '02 - Source'
+    runs-on: [ubuntu-22.04]
+    if: always()
+    needs: [clang_check]
+    defaults:
+      run:
+        shell: bash
+        working-directory: 'obs-studio'
+    steps:
+      - name: 'Checkout'
+        uses: actions/checkout@v3
+        with:
+          submodules: 'recursive'
+          path: 'obs-studio'
+          fetch-depth: 0
+
+      - name: 'Setup environment'
+        id: setup
+        run: |
+          echo "::set-output name=commitHash::$(git rev-parse --short=9 HEAD)"
+
+      - name: 'Create source artifact'
+        if: ${{ success() && github.event_name != 'pull_request' }}
+        run: |
+          outdir=$(pwd)/build
+          mkdir -p ${outdir}
+          prefix=obs-studio-${{ steps.setup.outputs.commitHash }}
+          outfile=${prefix}.tar
+          rm -f ${outfile}
+          # exclude repository configuration files
+          find .git -type d -name info -exec touch {}/attributes ";"
+          echo ".git* export-ignore" | find .git -type f -path "*/info/attributes" -exec tee -a {} +
+          # create a tarball with all the submodules
+          # this uses uncompressed 'tar', so we can append the archives
+          git archive --format tar -o ${outdir}/${outfile}  --prefix=$prefix/ HEAD
+          PREFIX=${prefix} OUTFILE=${outdir}/${outfile} git submodule --quiet foreach --recursive 'git archive --format tar -o submodule.tar --prefix=$PREFIX/$displaypath/ $sha1; tar -Af ${OUTFILE} submodule.tar; rm -f submodule.tar'
```

Comment:
This doesn't take a possible error in each previous step into account and will lead to broken archives. `tar` should only be invoked if the submodule fetch was successful.

Also `rm` will throw an error, if the tar file is not created. One could argue that the entire command will abort the workflow because of that, but you should abort early, especially as consecutive commands implicitly depend on prior commands to be successful.

Reply:
i'm not sure i can follow.
the code is written on the assumption that the sCript is executed with `set -e` (as is the case on CI-systems).

so the only problem i see is that within the `foreach` loop, `tar -Af` is executed even if `tar -o submodule.tar` failed...
i'll fix that.

in the CI context, i'll remove the deletion of `submodule.tar` (as it happens in a throwaway container anyhow)

Reply:
You execute a command and rely on its success for a consecutive command. Even without `set -e` this should use `&&` to ensure that consecutive commands only run on success to avoid relying on functionality you cannot control. Also the shell grammar allows you to be explicit about it.

Also have you checked if git aborts if an iteration of a `foreach` command has a non-zero return?

Reply:
i did check `git foreach` (admittedly, after your critique), and it indeed aborts if a sub-command returns non-0